app/javascript/utilities/viewport.js

```diff
@@ -22,3 +22,37 @@ export function isInViewport(element, offsetTop = 0) {
     boundingRect.right <= clientWidth
   );
 }
+
+/**
+ * Checks if any element is partially visible in the viewport Three cases to consider an element to be partially visible
+ * 1) The element's top portion lies above the Viewport
+ * 2) The element's bottom portion lies below the viewport
+ * 3) The element covers the entire viewport
+ * @example
+ * const element = document.getElementById('element');
+ * isPartiallyVisibleInViewport(element); // true or false
+ * @param {object} element - The HTML element to check
+ * @param {number} [offsetTop=0] - Part of the screen to ignore counting from the top
+ *
+ * @returns {boolean} isInViewport - true if the element is visible in the viewport
+ **/
+export function isPartiallyVisibleInViewport(element, offsetTop = 0) {
+  const boundingRect = element.getBoundingClientRect();
+  const clientHeight =
+    window.innerHeight || document.documentElement.clientHeight;
+  const clientWidth = window.innerWidth || document.documentElement.clientWidth;
+  return (
```

Comment:
There is some stuff that can be reused in all those conditions, and it will help make the condition be more readable overall.

I would consider creating variables for each of the bounding rectangle's sides that defines is that particular side is visible in the viewport.

For example:
```js
const topIsInViewport = boundingRect.top <= clientHeight && boundingRect.top >= offsetTop;
const rightIsInViewport = boundingRect.right >= 0 && boundingRect.right <= clientWidth;
const bottomIsInViewport = boundingRect.bottom >= offsetTop && boundingRect.bottom <= clientHeight;
const leftIsInViewport = boundingRect.left <= clientWidth && boundingRect.left >= 0;
```

That way, you can use those variables in any combination in order to get the value you want.

For example, if you want to know if an element is 100% in the viewport, then all of those variables need to be true:
```js
return (
  topIsInViewport &&
  bottomIsInViewport &&
  leftIsInViewport &&
  rightIsInViewport
);
```

And if you want to know if an element is partially in the viewport, it needs to satisfy the following 2 conditions:
- its top OR its bottom is visible
- its left OR its right is visible

That way, you can use the following condition:
```js
return (topIsInViewport || bottomIsInViewport) && (leftIsInViewport || rightIsInViewport);
```

With all of that said, I think we can merge our 2 viewport functions. We can pass it an optional parameter that defines if we want the element to be partially visible or completely visible. We can then use the corresponding condition in our return statement.

Reply:
Thank you so much for taking the time to review it and giving valuable suggestions @reobin. I have made the changes as per your suggestion. 

And also there was one more following case apart from the above-mentioned ones,

```
const topIsOutOfViewport = boundingRect.top <= offsetTop;
const bottomIsOutOfViewport = boundingRect.bottom >= clientHeight;
const elementSpansEntireViewport =
    topIsOutOfViewport && bottomIsOutOfViewport;
```

The _elementSpansEntireViewport_ should also be part of the partially visible section check as below

```
return (topIsInViewport || bottomIsInViewport || elementSpansEntireViewport) &&
        (leftIsInViewport || rightIsInViewport)
```

I have added the changes and pushed them. Review it when you find the time and let me know. Cheers!

Reply:
Hey @reobin!

I have made the changes you suggested. Can you go over it when you find the time and let me know if it is fine?

Reply:
Hey @gowtham2095!

Sorry for the delay on the review! I have been very busy with personal life stuff. I'll let the Forem team take care of this. Good work!

Reply:
Hi @reobin,

I understand. Not a problem. Take care!